tests/drivers/build_all/sensor/src/generic_test.c

```diff
@@ -176,36 +177,71 @@ static void run_generic_test(const struct device *dev)
 		/* Release the CQE */
 		rtio_cqe_release(&sensor_read_rtio_ctx, cqe);
 
-		enum sensor_channel channel;
-		sensor_frame_iterator_t fit = {0};
-		sensor_channel_iterator_t cit = {0};
 		const struct sensor_decoder_api *decoder;
-		int8_t shift;
-		q31_t q;
+		union sensor_data_union decoded_data;
 
 		zassert_ok(sensor_get_decoder(dev, &decoder));
 
-		/* Decode the buffer and verify all channels */
-		while (decoder->decode(buf, &fit, &cit, &channel, &q, 1) > 0) {
-			zassert_true(channel_table[channel].supported);
-			zassert_false(channel_table[channel].received);
-			channel_table[channel].received = true;
+		/* Loop through each channel */
+		for (int ch = 0; ch < ARRAY_SIZE(channel_table); ch++) {
+			if (!channel_table[ch].supported) {
+				continue;
+			}
 
-			zassert_ok(decoder->get_shift(buf, channel, &shift));
+			sensor_frame_iterator_t fit = {0};
+
+			rv = decoder->decode(buf, ch, 0, &fit, 1, &decoded_data);
+			zassert_equal(1, rv, "Could not decode (error %d, ch %d, iteration %d/%d)",
+				      rv, ch, iteration + 1,
+				      CONFIG_GENERIC_SENSOR_TEST_NUM_EXPECTED_VALS);
+
+			channel_table[ch].received = true;
+
+			/* Retrieve the actual value */
+			q31_t q;
+			int8_t shift;
+
+			switch (ch) {
+			/* Special handling to break out triplet samples. */
```

Comment:
Can we just skip these? I'm assuming that `MAGN_X` will be caught by `MAGN_XYZ` anyway no? I think we'll likely end up deprecating the fan-out channels.

Reply:
The emulator treats the channels separately to allow for the possibility of X, Y, Z having different reported ranges, in case the sensor doesn't support the same range on all axes (not sure if that exists in the wild) or if the user wants to set different per-axis offsets (more likely, but not possible for now until the backend grows the ability to modify attributes). We would have to give that up or modify the channel discovery and backend emul API to support setting a 3-tuple of expected ranges. Either will result in at least some refactoring and delay this PR.